autorecsys/pipeline/preprocessor.py

```diff
@@ -78,6 +103,216 @@ def preprocessing(self, **kwargs):
+        gc.collect()
+
+        # Step 4: Format data and obtain hash statistics.
+        array_data = np.concatenate((np.asarray(label_ll).T, np.asarray(numer_ll).T, np.asarray(categ_ll).T), axis=1)
+        del label_ll  # release memory
+        del numer_ll
+        del categ_ll
+        gc.collect()
+
+        self.pd_data = pd.DataFrame(array_data, columns=self.used_columns_names)
+        self.hash_sizes = [self.pd_data[categ_name].nunique() for categ_name in self.categ_names]
+        del array_data  # release memory
+        gc.collect()
+
+        for col_name, dtype in self.dtype_dict.items():
+            self.pd_data[col_name] = self.pd_data[col_name].astype(dtype)
+        # Save transformed data.
+        save_pickle(self.transformed_data_path, self.pd_data)
+
+    def scale_numerical_data(self):
+
+        # TODO: Designated transformation functions to specified place.
+        def scale_by_natural_log(num):
+            # TODO: 1) Explain why the conditional statement makes exception for numbers like 1, where 1>ln(1)**2
+            if num > 2:
+                num = math.log(float(num))**2
+            return num
+
+        for numer_name in self.numer_names:
+            self.pd_data[numer_name] = self.pd_data[numer_name].map(scale_by_natural_log)
+
+    def split_data(self, X, y, train_size, validate_size, random_state=42):
+        """
+
+        :param X: numpy array
+        :param y: numpy array
+        :param train_size:
+        :param validate_size:
+        :param random_state:
+        :return:
+        """
+        # Step 1: Obtain shuffled data indices.
+        data_size = X.shape[0]
+        np.random.seed(random_state)
+        shuffled_indices = np.random.permutation(range(data_size))
+
+        # Step 2: Determine data split positions.
+        train_end = int(train_size * data_size)
+        validate_end = train_end + int(validate_size * data_size)
+
+        # Step 3: Split shuffled data.
+        train_X = X[shuffled_indices][:train_end]
+        train_y = y[shuffled_indices][:train_end]
+        validate_X = X[shuffled_indices][train_end:validate_end]
+        validate_y = y[shuffled_indices][train_end:validate_end]
+        test_X = X[shuffled_indices][validate_end:]
+        test_y = y[shuffled_indices][validate_end:]
+
+        return train_X, validate_X, test_X, train_y, validate_y, test_y
+
+    def preprocessing(self, train_size, validate_size, random_state=42):
+        self.X = self.pd_data.iloc[:, 1:].values
+        self.y = self.pd_data.iloc[:, [0]].values
+
+        train_X, validate_X, test_X, self.train_y, self.validate_y, self.test_y = self.split_data(
+            self.X, self.y, train_size=train_size, validate_size=validate_size, random_state=random_state)
+
+        # Reformat numerical and categorical data.
+        self.train_X = [train_X[:, :self.numer_num], train_X[:, self.numer_num:]]
+        del train_X  # release memory
+        gc.collect()
```

Comment:
it is necessary to use both del and gc.collect() multiple times? can we do it one time?

Reply:
The non-consecutive/line-after-line free memory statements exist because Criteo (and any potential dataset) has large space cost and we are allocating more memory between these free statements.